marker/processors/high_quality_text.py

```diff
@@ -0,0 +1,207 @@
+
+        text_lines = block.contained_blocks(document, (BlockTypes.Line,))
+        extracted_lines = [line.formatted_text(document) for line in text_lines]
+
+        prompt = self.gemini_rewriting_prompt + '```json\n`' + json.dumps({"extracted_lines": extracted_lines}, indent=2) + '`\n```\n'
+        image = self.extract_image(page, block)
+        response_schema = content.Schema(
+            type=content.Type.OBJECT,
+            enum=[],
+            required=["corrected_lines"],
+            properties={
+                "corrected_lines": content.Schema(
+                    type=content.Type.ARRAY,
+                    items=content.Schema(
+                        type=content.Type.STRING,
+                    ),
+                )
+            },
+        )
+
+        response = self.generate(prompt, image, response_schema)
+        corrected_lines = []
+        if response and "corrected_lines" in response:
+            corrected_lines = response["corrected_lines"]
+
+        if corrected_lines and len(corrected_lines) == len(extracted_lines):
+            for text_line, corrected_text in zip(text_lines, corrected_lines):
+                span_block = page.add_full_block(
+                    SpanClass(
+                        polygon=text_line.polygon,
+                        text=corrected_text + "\n",
+                        font='Unknown',
+                        font_weight=0,
+                        font_size=0,
+                        minimum_position=0,
+                        maximum_position=0,
+                        formats=['plain', 'math'],
+                        page_id=text_line.page_id,
+                        text_extraction_method="gemini",
+                    )
+                )
+                text_line.structure = [span_block.id]
+
+    def extract_image(self, page: PageGroup, image_block: Block, expand: float = 0.01):
+        page_img = page.lowres_image
+        image_box = image_block.polygon\
+            .rescale(page.polygon.size, page_img.size)\
+            .expand(expand, expand)
+        cropped = page_img.crop(image_box.bbox)
+        return cropped
+
+    def generate(self, prompt: str, image: PIL.Image.Image, response_schema: content.Schema):
+        while True:
+            try:
+                responses = self.model.generate_content(
+                    [prompt, image],
+                    stream=False,
+                    generation_config={
+                        "temperature": 0,
+                        "response_schema": response_schema,
+                        "response_mime_type": "application/json",
+                    }
+                )
+                output = responses.candidates[0].content.parts[0].text
+                return json.loads(output)
+
+            except ResourceExhausted as e:
+                print(f"ResourceExhausted: {e}")
+                time.sleep(tries * 2)
+                tries += 1
```

Comment:
tries doesn't seem to be defined.  There's also no max_tries to stop after some time

Reply:
fixed it!